File: DATA STRUCTURES AND ALGORITHMS/AT/WEEK2/modulo.py
def fibonacci(n):
    f = list()
    f = [0,1]
    for i in range(2,n+1):
        f.append(f[i-1] + f[i-2])
    return f[n]

def pinaso_period(m):
    previous,current = 0,1
    p = 0
    while(True):
        p += 1
        previous,current = current, (previous + current) % m
        if (previous == 0 and current == 1):
            break
    return p

def modulo(n,m):
    pp = pinaso_period(m)
    r = n % pp
    t = fibonacci(r)
    return t % m

n,m =  (input()).split(" ")
ans = modulo(int(n),int(m))
print (ans)

File: DATA STRUCTURES AND ALGORITHMS/AT/WEEK2/modulo.c
#include <stdio.h>

unsigned long long int fibonacci(int n){
    unsigned long long int f[n+1];
	int i;
	f[0] = 0;
	f[1] = 1;
	for (i=0;i<=n;i++){
	   if (i >= 2)
        f[i] = f[i-2] + f[i-1];
      }
    printf("%llu",f[n]);
	return f[n];
}

int pisano_period(int m){
    int previous = 0,current = 1;
    int pp = 0;
    while (1){
    	pp++;
    	int temp = current;
    	current = (previous + current) % m;
    	previous = temp;
    	if (previous == 0 && current == 1){
    		break;
    	}
    }
    return pp;
}

unsigned long long int modulo(unsigned long long int n, int m){
   int p = pisano_period(m);
   int r = n % p;
   unsigned long long int f = fibonacci(r);
   return f % m;
}

int main(){
	int m;
	unsigned long long int n;
	scanf("%llu %d",&n,&m);
	printf("\n%llu",modulo(n,m));
}

